crates/sui-graphql-rpc/src/types/coin_metadata.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{context_data::db_data_provider::PgManager, error::Error};
+
+use super::big_int::BigInt;
+use super::move_object::MoveObject;
+use async_graphql::*;
+
+use sui_json_rpc::coin_api::parse_to_struct_tag;
+use sui_types::balance::Supply;
+use sui_types::coin::CoinMetadata as SuiCoinMetadata;
+use sui_types::gas_coin::{GAS, TOTAL_SUPPLY_SUI};
+
+pub(crate) struct CoinMetadata {
+    pub super_: MoveObject,
+    pub native: SuiCoinMetadata,
+}
+
+#[Object]
+impl CoinMetadata {
+    /// Convert the coin metadata object into a Move object
+    async fn as_move_object(&self) -> &MoveObject {
+        &self.super_
+    }
+
+    async fn decimals(&self) -> Option<u8> {
+        Some(self.native.decimals)
+    }
+
+    async fn name(&self) -> Option<&String> {
+        Some(&self.native.name)
+    }
+
+    async fn symbol(&self) -> Option<&String> {
+        Some(&self.native.symbol)
+    }
+
+    async fn description(&self) -> Option<&String> {
+        Some(&self.native.description)
+    }
+
+    async fn icon_url(&self) -> Option<&String> {
+        self.native.icon_url.as_ref()
+    }
+
+    async fn supply(&self, ctx: &Context<'_>) -> Result<Option<BigInt>> {
+        let coin_type = &self.super_.native.type_().type_params()[0];
```

Comment:
We shouldn't use `[0]` here, because it can panic -- can you switch this to a non-panicking access method, like `first()`?